File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    TreeNode lcaNode;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        lcaNode = root;
        lca(root , p , q);
        return lcaNode;
    }
    public void lca(TreeNode node , TreeNode p , TreeNode q){
        if(node.val>p.val && node.val>q.val)
            lca(node.left , p , q);
        else if(node.val<p.val && node.val<q.val)
            lca(node.right , p , q);
        else
            lcaNode = node;
        
    }
};

File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* P, TreeNode* Q) {
                  if(root == NULL){
            return NULL;
        }	
    
        if(root->val < P->val && root->val < Q->val){
            return lowestCommonAncestor(root->right,P,Q);
        }
       else if(root->val > P->val && root->val > Q->val){
            return lowestCommonAncestor(root->left,P,Q);
       }
        
       return root;
    }
};
